src/lockTracer.cpp

```diff
@@ -116,7 +119,11 @@ void JNICALL LockTracer::MonitorContendedEntered(jvmtiEnv* jvmti, JNIEnv* env, j
     }
 
     // Time is meaningless if lock attempt has started before profiling
-    if (_enabled && entered_time - enter_time >= _threshold && enter_time >= _start_time) {
+    const jlong duration = entered_time - enter_time;
+    unsigned long tid = (unsigned long)syscall(SYS_gettid);
+
+    // _threshold is used both as a duration threshold and a bucket interval. When the counter overflows _threshold, the event is sampled.
+    if (_enabled && duration >= _threshold && enter_time >= _start_time && updateCounter(_total_duration, duration, _threshold)) {
```

Comment:
`duration >= _threshold` condition is exactly how the current algorithm works.
The idea of a counter was to _replace_ this condition.